Lua settings improvements

PilzAdam opened the pull request:
Changes:
- Key settings are added to `minetest.conf.example`
- Added missing key and noclip settings to `settingtypes.txt
- Expanded description on key settings
- Fixed parallax typos
- `flags` now fully supported with a list of possible values
- Added input validation for `int`, `float` and `flags` types (type safety and min / max checks)
- Restructured graphics section
- Additional `settingtypes.txt` files are passed from games and mods root folders; they are displayed in special categories.
- Added description for format of `noise_params` type

PilzAdam committed:
Improve Lua settings menu

* Add key settings to setting table and ignore them later
  This way they are added to the auto-generated minetest.conf.example
* Add flags type
* Add input validation for int, float and flags
* Break in-game graphic settings into multiple sections
* Parse settingtpes.txt in mods and games
* Improve description for a lot of settings
* Fix typos and wording in settingtypes.txt
* Convert language setting to an enum

kilbith commented:
@PilzAdam L.267, please change "View" to "Viewing" like the setting above.

kilbith commented:
@PilzAdam You've cancelled (at least) 9d21fa3 with 5675486.

PilzAdam commented:
@kilbith Thanks for spotting that; I check everything else now.

PilzAdam commented:
Everything important is now done here, except the support for grouped noise param settings, but this is due to a missing Lua API functionality for grouped settings (or structs).

The only things missing from the list in #3266 are fancier input dialogs for certain types, but that's not so high priority.

I think this can be merged now.

est31 commented on builtin/settingtypes.txt:
```diff
@@ -29,6 +31,11 @@
 #   * key:
 #            - default
+#   * flags:
+#            Flags are always comman seperated without spaces.
```
comman?

PilzAdam replied:
Typo; will fix.

est31 commented on builtin/settingtypes.txt:
```diff
@@ -430,40 +583,41 @@ show_debug (Show debug info) bool false
 bind_address (Bind address) string
 
-#    Set to true to disallow old clients from connecting.
+#    Set to true to disallow old, possibly not fully compatible clients from connecting.
```
What does this have to do with "full compatibility"? If we support a network protocol version, we _are_ fully compatible to it.

PilzAdam replied:
Maybe not from the server admin's perspective. Older clients may be fully compatible from a network point of view, but they may not support all the new features that the server expects.

est31 commented on builtin/settingtypes.txt:
```diff
@@ -619,20 +770,21 @@ profiler_print_interval (Profiling print interval) int 0
 dedicated_server_step (Dedicated server step) float 0.1
 
-#    Can be set to true to disable shutting down on invalid world data.
+#    Can be enabled to disable shutting down on invalid world data.
```
Perhaps it would be easier to understand if phrased "If enabled, invalid world data won't cause the server to shut down"?

PilzAdam replied:
Will fix.

est31 commented on builtin/settingtypes.txt:
```diff
@@ -789,36 +938,38 @@ name (Player name) string
+#    -    action
+#    -    info
+#    -    verbose
+debug_log_level (Debug log level) enum action ,warning,action,info,verbose
```
why the leading comma?

PilzAdam replied:
This adds an empty entry to the enum

> `<nothing> (no logging)`

est31 replied:
I see.

est31 commented on builtin/mainmenu/tab_settings.lua:
```diff
@@ -190,22 +234,77 @@ local function parse_config_file()
+				if not games_category_initialized then
+					table.insert(settings, {
+						name = "Games",
+						translation = fgettext_ne("Games"), -- not used, but needed for xgettext
```
Perhaps make it more obvious that it really isn't needed, eg by making `translation` a local variable?

PilzAdam replied:
Will fix.

est31 commented:
Really great idea to scan for `settingstypes.txt` in the mods and game directories as well.

est31 commented:
For modders and game creators: should we add documentation for how to automatically generate minetest.conf.example for mods/games from settingstypes.txt?

PilzAdam commented:
@est31 addressed all your remarks.

If we want to expose the generation of `minetest.conf.example` then it should be done with a command line option or an extra button in the GUI. Editing `builtin/mainmenu/tab_settings.lua` is not an option for end-users (on some distros they may not be able to edit this file, since it's installed in the system-wide part).

I would personally prefer a way to pass command-line options to the main menu (maybe with registering them first, so they can be printed in `--help`), but that may be too complicated for this use-case.

rubenwardy commented on builtin/settingtypes.txt:
```diff
@@ -49,12 +58,18 @@
+#    This requires the "noclip" privilege on the server.
+noclip (Noclip) bool false
+
 #    Smooths camera when moving and looking arround.
```
"arround"

rubenwardy commented on builtin/settingtypes.txt:
```diff
@@ -281,63 +308,181 @@ texture_clean_transparent (Clean transparent textures) bool false
+#    Horizontal initial window size.
+screenH (Screen height) int 600
+
+#    Fullscreen mode.
```
I'd argue fullscreen isn't advanced.

rubenwardy commented on builtin/settingtypes.txt:
```diff
@@ -490,38 +648,35 @@ max_packets_per_iteration (Max. packets per iteration) int 1024
 
-#    Maximum number of players connected simultaneously.
+#    Maximum number of players that can connect simultaneously.
 max_users (Maximum users) int 15
```
I'd argue this is a server setting, rather than a game setting: same for motd.

"game" for me means settings related to gameplay.

rubenwardy commented on builtin/settingtypes.txt:
```diff
@@ -660,22 +816,21 @@ max_block_generate_distance (Max block generate distance) int 6
-#    Currently supported: trees, caves, flat, dungeons, light.
 #    Flags that are not specified in the flag string are not modified from the default.
-#    To explicitly turn off a flag, prepend "no" to the beginning, e.g. nolight.
+#    Flags starting with "no" are used to explicitly disable them.
```
The example was useful.

rubenwardy commented on builtin/settingtypes.txt:
```diff
@@ -780,45 +931,49 @@ secure.enable_security (Enable mod security) bool false
-language (Language) string
+#    Set the language. Leave empty to use the system language.
+#    A restart is required after changing this.
+language (Language) enum   ,be,cs,da,de,eo,es,et,fr,hu,id,it,ja,jbo,ko,ky,lt,nb,nl,pl,pt,pt_BR,ro,ru,tr,uk,zh_CN,zh_TW
```
What about en?

rubenwardy commented on src/settings_translation_file.cpp:
```diff
@@ -22,42 +24,168 @@ fake_function() {
+	gettext("Camera update toggle key");
+	gettext("Key for toggling the camrea update. Only used for development\nSee http://irrlicht.sourceforge.net/docu/namespaceirr.html#a54da2a0e231901735e3da1b0edf72eb3");
+	gettext("Debug info toggle key");
+	gettext("Key for toggling the display of debug info.\nSee http://irrlicht.sourceforge.net/docu/namespaceirr.html#a54da2a0e231901735e3da1b0edf72eb3");
```
as you can't click links, you should make a tinyurl style shortcut, or a wiki link.

PilzAdam replied:
These settings are not visible in GUI, only in the generated mintest.conf.example.

paramat commented:
New notes for map_generation_limit for settingtypes.txt and minetest.conf.example.
The unnecessary 'please note' line can be removed and the notes formatted as elsewhere in these files.

Map generation is limited to this distance in nodes from (0, 0, 0), along all axes in positive and negative directions. 
Value is limited to 31000.
By default mapgen generates in chunks of 80x80x80 nodes.
By default these chunks are offset (-32, -32, -32) nodes from the origin.
Only chunks within the map_generation_limit are generated.

PilzAdam commented:
Thanks for all your suggestions on `settingtypes.txt`, but I'd like to first merge this PR, since it contains important functionality. Improving `settingtypes.txt` will probably take a long time and should be subject of future PRs.

paramat commented:
Cool. For the moment don't worry too much about the setting of noise parameters, as this is advanced stuff that can always be done using minetest.conf.example.

est31 commented:
Don't edit that file anymore. If you want to change something in minetest.conf.example, edit settingstypes.txt instead, and @PilzAdam will generate minetest.conf.example with automated tools.
